4.next - Use new fixtures for tests

markstory opened the pull request:
Extract schema from the existing core fixtures into a new schema file that can be used with data-only fixtures. I had to make a few API changes to the new fixtures as I hadn't wired up row loading yet.

I don't think we can safely remove schema from any of the core fixtures as they are likely referenced in plugin builds and those plugins will lag behind in adopting new fixtures.

Part of #15308

markstory committed:
Fix data-only fixtures never actually loading fixture data.

Replace `fixturize()` with `setupTest()` that is only implemented by
data-only fixtures. This makes it easier to reduce reliance on PHPUnit
integrations and connect fixtures to test setup.

markstory committed:
Add first draft of schema data

I generated this from the existing fixture classes using reflection.

markstory committed:
Fix tests run in isolation failing due to a missing fixtureManager

markstory committed:
Update phpunit.xml to use data-only fixtures.

othercorey commented:
Is the schema.php array in any specific order?

markstory commented:
> Is the schema.php array in any specific order?

Yes and no. The tables are generally not ordered except where foreign keys are involved.

> Is it possible to load from multiple files or do teams have to deal with maintaining a single large array?

Ideally no users ever have to interact with this format. This format is only intended to be used for core tests, as we can't require phinx or migrations due to cyclic dependencies on cake core.

Instead developers would run migrations for their application in their `tests/bootstrap.php` The start of that work is in cakephp/migrations#488. Alternative forms for loading schema include loading one or more files with the required SQL to create the schema, or users can roll their own schema generation if they have externally managed SQL via another migration tool.

markstory committed:
Remove invalid annotation.

markstory committed:
Don't use schema reflection on throwaway schema

Describing a table, and parsing the metadata is expensive when we
are going to throw the results away. Creating a `TableSchema` object is
good enough to create a truncate statement. This helps improve
performance in MySQL and Postgres.

markstory committed:
Make truncation more efficient

Only truncate tables that had fixture data loaded. Truncating all
fixtures gets really slow as we churn the entire schema. This also
better emulates the behavior of the current fixture manager.

markstory committed:
Fix test.

This test was passing by luck before, as it relied on double loading
a fixture to trigger a schema error. Now we start off with schema in the
database and don't need to double load the fixture.

markstory commented on src/TestSuite/Fixture/TruncationStrategy.php:
```diff
@@ -96,14 +98,13 @@ public function beforeTest(string $test): void
             }
+            $schema = $db->getSchemaCollection();
+            $tables = $schema->listTables();
+            $tables = array_intersect($schema->listTables(), $fixtures->lastInserted());
```
I'm not in love with the `lastInserted()` approach and would appreciate any ideas on a cleaner solution.

The problem I was aiming to solve was that truncating the 54 fixture tables we have was really slow. The current fixture system skips some of this work by only truncating tables that were used in the current test.

othercorey replied:
> The current fixture system skips some of this work by only truncating tables that were used in the current test.

Our internal fixtures don't have foreign keys setup most of the time. Do we want to support fixtures with foreign keys and truncating subsets of the tables?

markstory committed:
Separate fixture usage to preserve constraints

Using the same fixtures in the FixtureManager test and TableSchema test
causes constraints to be dropped. I'm not able to reproduce the
TableSchema test failure locally with postgres 9.6

othercorey commented:
> Ideally no users ever have to interact with this format. This format is only intended to be used for core tests, as we can't require phinx or migrations due to cyclic dependencies on cake core.

Should we enforce that only core uses it?

markstory commented:
> Should we enforce that only core uses it?

It is marked as `@internal` and won't be included in the docs.

othercorey commented on src/TestSuite/Fixture/FixtureLoader.php:
```diff
@@ -79,17 +79,26 @@ abstract public function loadSingle(
+     *
+     * @return string[]
+     */
+    abstract public function lastInserted(): array;
```
This name doesn't make much sense. Fixtures aren't "inserted" and "last" doesn't really mean anything either.

I know this was a hack/temp solution but maybe clarifying this will help.

markstory replied:
I agree the name is clunky. I am planning to make this method protected when I move state management. I didn't want to increase the scope of this change too much as I am still battling with the various database differences.

markstory committed:
Debug sqlserver

markstory committed:
Drop constraints during reset and reflect schema again

SQLServer cannot drop tables if they are referenced in constraints.
Before trying to drop tables we need to remove constraints. Furthermore,
SQLServer needs to have the full schema reflection in order to reset
sequences when truncating. This schema can be cached though as we don't
want to reflect schema on each test.

markstory committed:
Fix type error.

markstory committed:
See all SQLServer errors.

markstory committed:
Improve performance in tests.

The duplicate connection was causing truncation to run multiple times on
the same database.

markstory committed:
Fix test referencing a connection that was removed.

markstory committed:
Skip getting a table list when we didn't insert any fixtures.

We won't be doing any truncation so there is no point in finding out
which tables exist.

markstory committed:
Fix tests in SQLserver

Because we create all the schema ahead of time the `autoFixtures`
behavior we were relying on to lazily create the table once the type
mappings were in place does not work. Instead we can map the nvarchar
type which is SQLServer's text type.

markstory committed:
Fix more tests for SQLServer

markstory committed:
Fix more tests in SQLServer.

markstory committed:
Stabilize constraint column order.

markstory committed:
More SQLServer skips.

I can't get these tests to fail locally so they are flaky in SQLServer.

markstory committed:
Fix more sqlserver tests and phpcs

markstory commented:
Merging this so I can continue work on the next block of changes which is to make state reset part of the TestCase.
